packages/flutter/lib/src/widgets/snapshot_widget.dart

```diff
@@ -0,0 +1,269 @@
+      return;
+    }
+    _devicePixelRatio = value;
+    markNeedsPaint();
+  }
+
+  /// A controller that determines whether to paint the child normally or to
+  /// paint a snapshotted version of that child.
+  SnapshotWidgetController get controller => _controller;
+  SnapshotWidgetController _controller;
+  set controller(SnapshotWidgetController value) {
+    if (value == controller) {
+      return;
+    }
+    controller.removeListener(_onRasterValueChanged);
+    final bool oldValue = controller.enabled;
+    _controller = value;
+    if (attached) {
+      controller.addListener(_onRasterValueChanged);
+      if (oldValue != controller.enabled) {
+        _onRasterValueChanged();
+      }
+    }
+  }
+
+  /// How the snapshot widget will handle platform views in child layers.
+  SnapshotMode get mode => _mode;
+  SnapshotMode _mode;
+  set mode(SnapshotMode value) {
+    if (value == _mode) {
+      return;
+    }
+    _mode = value;
+    markNeedsPaint();
+  }
+
+  ui.Image? _childRaster;
+
+  @override
+  void attach(covariant PipelineOwner owner) {
+    controller.addListener(_onRasterValueChanged);
+    super.attach(owner);
+  }
+
+  @override
+  void detach() {
+    controller.removeListener(_onRasterValueChanged);
+    _childRaster?.dispose();
+    _childRaster = null;
+    super.detach();
+  }
+
+  @override
+  void dispose() {
+    controller.removeListener(_onRasterValueChanged);
+    _childRaster?.dispose();
+    _childRaster = null;
+    super.dispose();
+  }
+
+  void _onRasterValueChanged() {
+    _childRaster?.dispose();
+    _childRaster = null;
+    markNeedsPaint();
+  }
+
+  // Paint [child] with this painting context, then convert to a raster and detach all
+  // children from this layer.
+  ui.Image? _paintAndDetachToImage() {
+    final OffsetLayer offsetLayer = OffsetLayer();
+    final PaintingContext context = PaintingContext(offsetLayer, Offset.zero & size);
+    super.paint(context, Offset.zero);
+    // This ignore is here because this method is protected by the `PaintingContext`. Adding a new
+    // method that performs the work of `_paintAndDetachToImage` would avoid the need for this, but
+    // that would conflict with our goals of minimizing painting context.
+    // ignore: invalid_use_of_protected_member
+    context.stopRecordingIfNeeded();
+    if (mode != SnapshotMode.forced && !offsetLayer.supportsRasterization()) {
+      if (mode == SnapshotMode.normal) {
+        throw FlutterError('SnapshotWidget used with a child that contains a PlatformView.');
+      }
+      return null;
+    }
+    final ui.Image image = offsetLayer.toImageSync(Offset.zero & size, pixelRatio: devicePixelRatio);
+    offsetLayer.dispose();
+    return image;
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (size.isEmpty) {
+      _childRaster?.dispose();
+      _childRaster = null;
+      return;
+    }
+    if (controller.enabled) {
+      _childRaster ??= _paintAndDetachToImage();
```

Comment:
conceptually, when this returns null in one frame, do we have to try again every consecutive frame to snapshot it? Seems like in this case we are essentially doing `super.paint` twice for every frame?